File: code/49-Group Anagrams/49-Group Anagrams.py
# -*- coding: utf-8 -*
"""
LeetCode 49 - Group Anagrams

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
r"""
@author: Kai Chen
@file: 49-Group Anagrams.py
@time: 2020/7/10 15:59
@desc:
Given an array of strings, group anagrams together.

Example:

Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]
Note:

All inputs will be in lowercase.
The order of your output does not matter.
"""


class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        def cal(s):
            str = list(s)
            str.sort()
            return "".join(str)
        digitMap = {}
        for str in strs:
            digit = cal(str)
            if digit not in digitMap.keys():
                digitMap[digit] = [str]
            else:
                digitMap[digit].append(str)
        return list(digitMap.values())


if __name__ == '__main__':
    s = Solution()
    print(s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

File: code/49-Group Anagrams/49-Group Anagrams.c
/*
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

 

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
Example 2:

Input: strs = [""]
Output: [[""]]
Example 3:

Input: strs = ["a"]
Output: [["a"]]
 

Constraints:

1 <= strs.length <= 104
0 <= strs[i].length <= 100
strs[i] consists of lower-case English letters.
*/


/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */

int cmp(const void *a_, const void *b_){
    char *a = (char *)a_;
    char *b = (char *)b_;
    return *a - *b;
}


typedef struct {
    char * key;
    int idx;
    UT_hash_handle hh;
} HashItem;

char *** groupAnagrams(char ** strs, int strsSize, int* returnSize, int** returnColumnSizes){
    char *** res = (char***)malloc(sizeof(char**) * strsSize);
    *returnSize = 0;
    *returnColumnSizes = (int*)malloc(sizeof(int) * strsSize);
    HashItem *strMap = NULL;

    for (int i = 0; i < strsSize; i ++) {
        int len = strlen(strs[i]);
        char * tmp = (char*)malloc(sizeof(char) * (len + 1));
        strcpy(tmp, strs[i]);

        qsort(tmp, len, sizeof(char), cmp);

        HashItem *pEntry = NULL;
        HASH_FIND_STR(strMap, tmp, pEntry);
        if (pEntry == NULL) {
            pEntry = (HashItem*)malloc(sizeof(HashItem));
            pEntry->key = (char*)malloc(sizeof(char) * (len + 1));
            strcpy(pEntry->key, tmp);
            pEntry->idx = *returnSize;
            HASH_ADD_STR(strMap, key, pEntry);

            res[*returnSize] = (char**)malloc(sizeof(char*) * strsSize);
            res[*returnSize][0] = (char*)malloc(sizeof(char) * (len + 1));
            strcpy(res[*returnSize][0], strs[i]);
            (*returnColumnSizes)[*returnSize] = 1;
            (*returnSize) ++;
        } else {
            int row = pEntry->idx;
            int col = (*returnColumnSizes)[row];
            res[row][col] = (char*)malloc(sizeof(char) * (len + 1));
            strcpy(res[row][col], strs[i]);
            (*returnColumnSizes)[row] ++;
        }
    }
    return res;
}